### PR TITLE
Fast ruling

### LINKED ISSUE #4463: Implement "analyze project" API endpoint on bridge
- [x] generate default TS config
- [x] analyze yaml, html
- [x] tsconfig.json references

### PR DESCRIPTION
**Pull request by ilia-kebets-sonarsource**
Fixes https://github.com/SonarSource/SonarJS/issues/4461
Fixes #4463 
Fixes https://github.com/SonarSource/SonarJS/issues/4469

### CODE REVIEW

File: packages/jsts/src/analysis/projectAnalysis.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  rules: RuleConfig[];
+  environments: string[];
+  globals: string[];
+  baseDir: string;
+  tsConfigs?: string[];
+  exclusions?: string[];
+  isSonarlint?: boolean;
+};
+
+export const defaultEnvironments = [
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/analysis/projectAnalysis.ts:54]: This is currently needed, but it's a duplication from the Java plugin. TBD

File: packages/jsts/src/analysis/projectAnalyzer.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ * @returns the JavaScript / TypeScript project analysis output
+ */
+export async function analyzeProject(input: ProjectAnalysisInput): Promise<ProjectAnalysisOutput> {
+  const { rules, environments, globals, baseDir, exclusions = [] } = input;
+  const inputFilenames = Object.keys(input.files);
+  const pendingFiles: Set<string> = new Set(inputFilenames);
+  const watchProgram = input.isSonarlint || hasVueFile(inputFilenames);
+  initializeLinter(rules, environments, globals);
+  searchPackageJsonFiles(baseDir, exclusions);
+  searchTSConfigJsonFiles(baseDir, exclusions);
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/analysis/projectAnalyzer.ts:55]: merge these so we walk only once

File: packages/jsts/src/analysis/projectAnalyzer.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            fileType: files[filename].fileType,
+            programId,
+          },
+          files[filename].language ?? DEFAULT_LANGUAGE,
+        );
+        pendingFiles.delete(filename);
+      }
+    }
+    deleteProgram(programId);
+    if (!pendingFiles.size) {
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/analysis/projectAnalyzer.ts:101]:
```suggestion
    if (PendingFiles.size === 0) {
```

File: packages/jsts/tests/analysis/projectAnalyzer.test.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    rules: defaultRules,
+    environments: defaultEnvironments,
+    globals: defaultGlobals,
+    baseDir: fixtures,
+    files: filesDBtoFilesInput(files.db),
+  };
+}
+
+const fixtures = path.join(__dirname, 'fixtures');
+const files = new FileFinder(() => {});
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/tests/analysis/projectAnalyzer.test.ts:78]: as we will collect tsconfig and package.json, the change will require to implement a filter in this callback.

File: packages/jsts/src/analysis/projectAnalysis/projectAnalyzer.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      programsCreated: [],
+    },
+  };
+  if (!inputFilenames.length) {
+    return results;
+  }
+  const pendingFiles: Set<string> = new Set(inputFilenames);
+  const watchProgram = input.isSonarlint || hasVueFile(inputFilenames);
+  initializeLinter(rules, environments, globals);
+  searchTSConfigJsonAndPackageJsonFiles(baseDir, exclusions);
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/analysis/projectAnalysis/projectAnalyzer.ts:68]: are we not using the return value here?

File: packages/jsts/src/linter/wrapper.ts [hunk 6]
```diff
@@ -145,7 +145,6 @@ export class LinterWrapper {
     language: JsTsLanguage = 'js',
   ): LintingResult {
     const key: LinterConfigurationKey = { fileType, language };
-    debug(`Using linter configuration for ${JSON.stringify(key)}`);
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/linter/wrapper.ts:148]: why remove this one?

File: packages/jsts/src/program/tsconfigs/index.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    typeof maxFilesForTypeChecking === 'undefined'
+      ? DEFAULT_MAX_FILES_FOR_TYPE_CHECKING
+      : maxFilesForTypeChecking;
+  if (emptyTsConfigs && files.length < maxFiles) {
+    const tsConfig = sonarLint
+      ? createTSConfigFile(undefined, [baseDir + '/**/*'])
+      : createTSConfigFile(files);
+    const { filename } = await writeTSConfigFile(tsConfig);
+    yield filename;
+  }
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/program/tsconfigs/index.ts:73]: what is all that

File: packages/html/tests/analysis/analyzer.test.ts [hunk 8]
```diff
@@ -21,7 +21,9 @@ import { join } from 'path';
 import { embeddedInput } from '../../../jsts/tests/tools';
 import { parseHTML } from '../../src/parser';
 import { setContext } from '@sonar/shared';
-import { analyzeEmbedded, initializeLinter } from '@sonar/jsts';
+import { analyzeEmbedded } from '@sonar/jsts';
+// importing this from @sonar/jsts causes a resolution issue
+import { initializeLinter } from '../../../jsts/src/linter';
```
Comment (ilia-kebets-sonarsource) [inline packages/html/tests/analysis/analyzer.test.ts:26]: check what we do with that

**Comment by ilia-kebets-sonarsource** [comment]
coverage for the TS codebase in `packages/` was removed by these changes:
- https://github.com/SonarSource/SonarJS/pull/4489/commits/fb098a1a1ec27cc39880628a8b508decae1660b8
- https://github.com/SonarSource/SonarJS/pull/4489/commits/5ef008528595210130667e351fecbfc68d4f8884
which were part of this PR: https://github.com/SonarSource/SonarJS/pull/4489

We will fix it in a separate PR.

**Review by saberduck — APPROVED** [review]

**Pull request merged**